Modules/_sqlite/connection.c

```diff
@@ -440,40 +442,56 @@ remove_callbacks(sqlite3 *db)
 static void
 connection_close(pysqlite_Connection *self)
 {
-    if (self->db) {
-        if (self->autocommit == AUTOCOMMIT_DISABLED &&
-            !sqlite3_get_autocommit(self->db))
-        {
-            /* If close is implicitly called as a result of interpreter
-             * tear-down, we must not call back into Python. */
-            if (_Py_IsInterpreterFinalizing(PyInterpreterState_Get())) {
-                remove_callbacks(self->db);
-            }
-            (void)connection_exec_stmt(self, "ROLLBACK");
+    assert(self->db);
+    if (self->autocommit == AUTOCOMMIT_DISABLED &&
+        !sqlite3_get_autocommit(self->db))
+    {
+        /* If close is implicitly called as a result of interpreter
+         * tear-down, we must not call back into Python. */
+        if (_Py_IsInterpreterFinalizing(PyInterpreterState_Get())) {
```

Comment:
IMO you should move this logic into the caller which has the knowledge if the function was called explicitly or by the finalizer.

For example, add a "autocommit_remove_callbacks" parameter which would be 0 in the close() method, and set to _Py_IsInterpreterFinalizing(...) by the finalize function.

Reply:
Hm, yeah that may be nicer. I'll try that.

Reply:
I think that played out nice. Have a look in 8ebc8a73196a0715ed6b629d0e9721384d1e660f; thanks for the tip.